apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -0,0 +1,247 @@
+﻿// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Datastore.V1
+{
+    // Note: this file would be generated.
+
+    public partial class DatastoreDb
+    {
+        /// <summary>
+        /// Creates a new <see cref="DatastoreDbBuilder"/> instance. This method is primarily present for discoverability.
+        /// </summary>
+        /// <returns>A new <see cref="DatastoreDbBuilder"/> instance.</returns>
+        public DatastoreDbBuilder CreateBuilder() => new DatastoreDbBuilder();
+
+    }
+
+    /// <summary>
+    /// Builder class for <see cref="DatastoreClient"/> to provide simple configuration of credentials, endpoint etc.
+    /// </summary>
+    public sealed partial class DatastoreDbBuilder : ClientBuilderBase<DatastoreDb>
+    {
+        /// <summary>
+        /// The settings to use for RPCs, or null for the default settings.
+        /// </summary>
+        public DatastoreSettings Settings { get; set; }
+
+        private EmulatorDetection _emulatorDetection;
+
+        /// <summary>
+        /// Specifies how the builder responds to the presence of emulator environment variables as described
+        /// by https://cloud.google.com/datastore/docs/tools/datastore-emulator.
+        /// </summary>
+        /// <remarks>
+        /// This property defaults to <see cref="EmulatorDetection.None"/>, meaning that environment variables are
+        /// ignored.
+        /// </remarks>
+        public EmulatorDetection EmulatorDetection
+        {
+            get => _emulatorDetection;
+            set => _emulatorDetection = GaxPreconditions.CheckEnumValue(value, nameof(value));
```

Comment:
Is is correct to validate this here?
Isn't the idiom to only validate on build?

Reply:
Not for things that *can* be validated independently. If the value isn't a valid enum value, it can never be correct, so it's okay to validate here, IMO.